Formula/gstreamer.rb

```diff
@@ -1,71 +1,262 @@
+  end
+
+  head do
+    url "https://gitlab.freedesktop.org/gstreamer/gstreamer.git", branch: "main"
+
+    resource "rs" do
+      url "https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs.git", branch: "main"
+    end
   end
 
   depends_on "bison" => :build
+  depends_on "cargo-c" => :build
   depends_on "gobject-introspection" => :build
   depends_on "meson" => :build
+  depends_on "nasm" => :build
   depends_on "ninja" => :build
   depends_on "pkg-config" => :build
+  depends_on "rust" => :build
+  depends_on "yasm" => :build
+  depends_on "cairo"
+  depends_on "dav1d"
+  depends_on "faac"
+  depends_on "faad2"
+  depends_on "fdk-aac"
+  depends_on "ffmpeg"
+  depends_on "flac"
   depends_on "gettext"
   depends_on "glib"
+  depends_on "glib-networking"
+  depends_on "graphene"
+  depends_on "gtk+3"
+  depends_on "gtk4"
+  depends_on "jpeg-turbo"
+  depends_on "json-glib"
+  depends_on "lame"
+  depends_on "libogg"
+  depends_on "libpng"
+  depends_on "libpthread-stubs"
+  depends_on "libshout"
+  depends_on "libsoup"
+  depends_on "libusrsctp"
+  depends_on "libvorbis"
+  depends_on "libvpx"
+  depends_on "openssl@1.1"
+  depends_on "opus"
+  depends_on "orc"
+  depends_on "pango"
+  depends_on "pygobject3"
+  depends_on "python@3.11"
+  depends_on "rav1e"
+  depends_on "rtmpdump"
+  depends_on "speex"
+  depends_on "srtp"
+  depends_on "taglib"
+  depends_on "theora"
+  depends_on "x264"
+  depends_on "xz"
 
   uses_from_macos "flex" => :build
+  uses_from_macos "curl"
+
+  on_macos do
+    # musepack is not bottled on Linux
+    # https://github.com/Homebrew/homebrew-core/pull/92041
+    depends_on "musepack"
+  end
+
+  on_linux do
+    depends_on "freeglut"
+  end
+
+  def python3
+    which("python3.11")
+  end
+
+  # These paths used to live in various `gst-*` formulae.
+  link_overwrite "bin/gst-*", "lib/ligst*", "lib/libges*", "lib/girepository-1.0/Gst*-1.0.typelib"
+  link_overwrite "lib/girepository-1.0/GES-1.0.typelib", "lib/gst-validate-launcher/*", "lib/gstreamer-1.0/*"
+  link_overwrite "lib/pkgconfig/gst*.pc", "lib/python3.11/site-packages/gi/overrides/*", "include/gstreamer-1.0/*"
+  link_overwrite "share/gir-1.0/Gst*.gir", "share/gir-1.0/GES-1.0.gir", "share/gstreamer-1.0/*"
+  link_overwrite "share/locale/*/LC_MESSAGES/gst-*.mo", "share/man/man1/g*"
+
+  # Avoid overlinking of `gst-python` python extension module.
+  # https://gitlab.freedesktop.org/gstreamer/gst-python/-/merge_requests/41
+  # TODO: Migrate patch to gstreamer monorepo.
+  patch :DATA
 
   def install
-    # Ban trying to chown to root.
-    # https://bugzilla.gnome.org/show_bug.cgi?id=750367
-    args = std_meson_args + %w[
+    (buildpath/"subprojects/gst-plugins-rs").install resource("rs")
+    site_packages = Language::Python.site_packages(python3)
+    # To pass arguments to subprojects (e.g. `gst-editing-services`), use
+    #   -Dsubproject:option=value
+    args = %W[
+      -Dpython=enabled
+      -Dlibav=enabled
+      -Dlibnice=disabled
```

Comment:
libnice is actually a really important gstreamer plugin, since it's essential for webrtc support.

I am not sure how you can solve the circular dep though. Maybe you could move libnice into the monorepo too? Is there anything else in Homebrew that uses libnice?

Reply:
> Maybe you could move libnice into the monorepo too?

Yea, maybe. It'll make updating libnice really annoying, though. It's also not clear to me that you get the same libnice if you build it before Gstreamer rather than after, which might be an issue for users.

> Is there anything else in Homebrew that uses libnice?

Not currently, no.

Reply:
The monorepo will pull in libnice in the same way that it pulls in gst-plugins-rs (both are separate repos that are meson wraps), so it will be identical to building it separately. If nothing else is using libnice, then I suggest merging it into the monorepo too.

Note that libnice is maintained by Olivier, who is a gstreamer core maintainer, and I don't expect anyone outside of gstreamer to use the project.

If you really want to keep it separate, I'd recommend building it twice: once as its own project without gstreamer support, then again as part of the monorepo to build only the gstreamer plugin. That'll require non-trivial build system patching, but it can be done.

Reply:
It's not quite the same as gst-plugins-rs, though, which kind of supports updating simultaneously with gstreamer.

If we build libnice with gstreamer, then we need to rebuild all of gstreamer every time we want to update only libnice (if we even remember to update it after it becomes part of gstreamer).

Reply:
The libnice wrap file is continuously updated, so you'd get updates whenever you update gstreamer (stable releases for instance). We test the two as a unit in the monorepo and in cerbero.

Reply:
Only getting updates when gstreamer is updated is also not really desirable.

Reply:
fwiw, gst-plugins-rs is on a completely different release cycle than the rest of gstreamer, so you're already doing that for the rust plugins.

Reply:
Yes, but hacking up gst-*-rs so that it can work when updated independently isn't really something we want to do (though avoiding that introduced the need for other workarounds...). libnice seems to still be usable when updated independently.

Reply:
I would argue that if nothing else in Homebrew is using libnice, then it's not really usable in its current state since gstreamer can only use libnice if the gstreamer plugin is also available :slightly_smiling_face: 

Like I said, we can fix the circular dep by building only the gstreamer plugin as part of the monorepo build, do you want me to help you with that?

Reply:
:shrug: It has at least ~40k downloads in the past year, with zero complaints about it being unusable, so I'd hazard that some users are using it for _something_ even if it doesn't really work with gstreamer.

Let's leave it this way for the moment. If multiple users complain about not being able to use libnice, then we can look into including it in the monorepo build.

Reply:
Actually, we get all the complaints about that :) people complain that the webrtc plugin doesn't work in Homebrew, and we have to tell them to use the official binaries instead.

Gstreamer's webrtc support is probably the most widely-used feature out there right now (other than plain playback).

Reply:
Send them here the next time they come around then, and we'll take a look.

Reply:
Hi! Thanks for documenting the issue: upgrading to 1.22.2 effectively made `webrtcbin` no longer be listed as plugin. I understand the circular dependency issue, can some work be done to include `libnice` so that `webrtcbin` can be supported? It's not clear what the best approach is if `libnice` still needs to be a standalone.

Reply:
Did webrtc work for you before? @nirbheek tells me that it was basically broken previously.

Reply:
Up until 1.20.0 at least what it took was for gstreamer + gst-plugins-bad to be installed via brew in which case it was working yes.

Reply:
Earlier, all plugin registration was broken unless you set the right env vars, and linking was also broken. webrtcbin was not _specifically_ broken.

I can help you with a solution for the circular dep, just let me know if you need my help.

One way would be to allow gstnice (inside libnice) to be built against a system libnice. The plugin only uses the version string at build time, nothing else.

I can patch that and talk to Olivier about upstreaming it.

Reply:
> One way would be to allow gstnice (inside libnice) to be built against a system libnice. The plugin only uses the version string at build time, nothing else.

That seems like it could work.